app/src/main/kotlin/com/intive/picover/common/viewmodel/MVIViewModel.kt

```diff
@@ -1,51 +1,26 @@
 package com.intive.picover.common.viewmodel
 
-import androidx.compose.runtime.MutableState
-import androidx.compose.runtime.State
-import androidx.compose.runtime.mutableStateOf
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.viewModelScope
-import com.intive.picover.common.viewmodel.event.Event
-import com.intive.picover.common.viewmodel.sideeffect.SideEffect
-import com.intive.picover.common.viewmodel.state.MVIState
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
 import kotlinx.coroutines.launch
 
-abstract class MVIViewModel<S : MVIState, E : Event, SE : SideEffect>(initialState: S) : ViewModel() {
+abstract class MVIViewModel<S : Any, E : Any, SE : Any>(initialState: S) : ViewModel() {
 
 	@Suppress("ktlint:standard:property-naming")
-	protected val _state = mutableStateOf(initialState)
-	val state: State<S> = _state
+	protected val _state = MutableStateFlow(initialState)
+	val state = _state.asStateFlow()
 	private val _sideEffects = MutableSharedFlow<SE>()
 	val sideEffects: Flow<SE> = _sideEffects
-	private val events = MutableSharedFlow<E>()
 
-	init {
-		collectEvents()
-	}
-
-	protected fun MutableState<S>.update(copyBlock: S.() -> S) {
-		_state.value = value.copyBlock()
-	}
-
-	protected abstract fun handleEvent(event: E)
-
-	fun emitEvent(newEvent: E) {
-		viewModelScope.launch {
-			events.emit(newEvent)
-		}
-	}
+	abstract fun event(event: E)
```

Comment:
Storing a MutableSharedFlow for events and emitting every event in separate coroutine seems a bit overcomplicated, we can simply just call this function and handle it directly in view models inheriting from MVIViewModel

Reply:
I was thinking about the event, which could be sent after some delay, e.g. as a end result of some processing or waiting the request, when the app would go to the background in the meantime - but this shouldn't be an event but side effect then, am I right?

Reply:
And I really like the simplification 🍻

Reply:
Regarding sending the event after delay, I would need to see a specific case, to be honest it's quite confusing now with events and side effects 😄